olive/passes/onnx/nvmo_quantization.py

```diff
@@ -0,0 +1,129 @@
+        default_value=1,
+        description="Batch size for calibration.",
+    ),
+    "calib_size": PassConfigParam(
+        type_=int,
+        default_value=64,
+        description="Calibration data size.",
+    ),
+    "dataloader_func": PassConfigParam(
+        type_=Union[Callable, str],
+        category=ParamCategory.OBJECT,
+        required=True,
+        description="Function/function name to generate dataloader for calibration.",
+    ),
+    "dataloader_func_kwargs": PassConfigParam(
+        type_=Dict[str, Any],
+        description="Keyword arguments for dataloader_func.",
+    ),
+}
+
+
+class NVModelOptQuantization(Pass):
+    """Quantize ONNX model with Nvidia-ModelOpt."""
+
+    # set this to True if the pass has parameters that are functions or objects and the user script is required
+    # to import the module containing the function or object
+    _requires_user_script: bool = True
+
+    class Precision(str, Enum):
+        FP8 = "fp8"
+        INT8 = "int8"
+        INT4 = "int4"
+
+        def __str__(self) -> str:
+            return self.value
+
+    class Algorithm(str, Enum):
+        RTN = "RTN"
+        AWQ = "AWQ"
+
+        def __str__(self) -> str:
+            return str(self.value)
+
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        config = {
+            "precision": PassConfigParam(
+                type_=NVModelOptQuantization.Precision,
+                default_value="int4",
+                searchable_values=Categorical(["fp8", "int8", "int4"]),
+                description="NVModelOpt Quantization mode.",
+            ),
+            "algorithm": PassConfigParam(
+                type_=NVModelOptQuantization.Algorithm,
+                default_value="RTN",
+                searchable_values=Categorical(["RTN", "AWQ"]),
+                description="Algorithm of weight only quantization. Support 'RTN' and 'AWQ'.",
+            ),
+        }
+
+        config.update(deepcopy(_dataloader_config))
+        return config
+
+    def validate_search_point(
+        self, search_point: Dict[str, Any], accelerator_spec: AcceleratorSpec, with_fixed_value: bool = False
+    ) -> bool:
+        if with_fixed_value:
+            search_point = self.config_at_search_point(search_point or {})
+
+        if search_point["precision"] != NVModelOptQuantization.Precision.INT4 or search_point["algorithm"] not in [
+            NVModelOptQuantization.Algorithm.RTN,
+            NVModelOptQuantization.Algorithm.AWQ,
+        ]:
+            logger.error("Only INT4 quantization with RTN and AWQ algorithm is supported.")
+            return False
+
+        return True
+
+    def _run_for_config(
+        self, model: OliveModelHandler, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> OliveModelHandler:
+        try:
+            from modelopt.onnx.quantization.int4 import quantize_int4  # type: ignore[import]
```

Comment:
Not a breaking issue since onnx opset 21 is not currently supported by ort. But I think the quantizer probably needs to update the model's opset to 21 if it is not already so? Or require the model's opset to be 21? (olive provides a pass for that too https://github.com/microsoft/Olive/blob/3e5ce3f6907eed5063d937af756e9a214a9342eb/olive/passes/onnx/conversion.py#L579)

Otherwise, the model graph is technically invalid for lower opset version since int4 in dq and q ops were only added in opset version 21. https://github.com/onnx/onnx/releases/tag/v1.16.0

Reply:
Our tool does not care much about opset version either. Anything else before we can merge this PR?

Reply:
Sounds good to me. Everything looks good from my side. I will approve and merge once the CI passes. 

The only outstanding thing is the discussion about the deployment messaging but that can be updated in a follow up PR if necessary.

Reply:
The failing/cancelled test is not related to this PR. I will rerun it once the other tests complete.